src/ott/solvers/quadratic/gromov_wasserstein_lr.py

```diff
@@ -0,0 +1,880 @@
+    lse_mode: bool,
+    use_danskin: bool = False
+  ) -> "LRGWOutput":
+    del lse_mode
+    return self.set(reg_ot_cost=self.compute_reg_ot_cost(ot_prob, use_danskin))
+
+  def compute_reg_ot_cost(  # noqa: D102
+    self,
+    ot_prob: quadratic_problem.QuadraticProblem,
+    use_danskin: bool = False,
+  ) -> float:
+    return compute_reg_ot_cost(
+        self.q,
+        self.r,
+        self.g,
+        ot_prob,
+        epsilon=self.epsilon,
+        use_danskin=use_danskin
+    )
+
+  @property
+  def linear(self) -> bool:  # noqa: D102
+    return False
+
+  @property
+  def geom(self) -> geometry.Geometry:  # noqa: D102
+    """Linearized geometry."""
+    # TODO(michalk8): extract to a common function (also in other places)
+    inv_sqrt_g = jnp.sqrt(self._inv_g)
+    tmp1 = -4.0 * self.ot_prob.geom_xx.apply_cost(self.q, axis=1) * inv_sqrt_g
+    tmp2 = self.ot_prob.geom_yy.apply_cost(self.r, axis=1) * inv_sqrt_g
+    return low_rank.LRCGeometry(tmp1, tmp2)
+
+  @property
+  def a(self) -> jnp.ndarray:  # noqa: D102
+    return self.ot_prob.a
+
+  @property
+  def b(self) -> jnp.ndarray:  # noqa: D102
+    return self.ot_prob.b
+
+  @property
+  def linear_output(self) -> bool:  # noqa: D102
+    return False
+
+  @property
+  def converged(self) -> bool:  # noqa: D102
+    return jnp.logical_and(
+        jnp.any(self.costs == -1), jnp.all(jnp.isfinite(self.costs))
+    )
+
+  @property
+  def matrix(self) -> jnp.ndarray:
+    """Transport matrix if it can be instantiated."""
+    return (self.q * self._inv_g) @ self.r.T
+
+  def apply(self, inputs: jnp.ndarray, axis: int = 0) -> jnp.ndarray:
+    """Apply the transport to a array; axis=1 for its transpose."""
+    q, r = (self.q, self.r) if axis == 1 else (self.r, self.q)
+    # for `axis=0`: (batch, m), (m, r), (r,), (r, n)
+    return ((inputs @ r) * self._inv_g) @ q.T
+
+  def marginal(self, axis: int) -> jnp.ndarray:  # noqa: D102
+    length = self.q.shape[0] if axis == 0 else self.r.shape[0]
+    return self.apply(jnp.ones(length,), axis=axis)
+
+  def cost_at_geom(self, other_geom: geometry.Geometry) -> float:
+    """Return OT cost for current solution, evaluated at any cost matrix."""
+    return jnp.sum(self.q * other_geom.apply_cost(self.r, axis=1) * self._inv_g)
+
+  def transport_cost_at_geom(self, other_geom: geometry.Geometry) -> float:
+    """Return (by recomputing it) bare transport cost of current solution."""
+    return self.cost_at_geom(other_geom)
+
+  @property
+  def primal_cost(self) -> float:
+    """Return (by recomputing it) transport cost of current solution."""
+    return self.transport_cost_at_geom(other_geom=self.geom)
+
+  @property
+  def transport_mass(self) -> float:
+    """Sum of transport matrix."""
+    return self.marginal(0).sum()
+
+  @property
+  def _inv_g(self) -> jnp.ndarray:
+    return 1.0 / self.g
+
+
+@jax.tree_util.register_pytree_node_class
+class LRGromovWasserstein(sinkhorn.Sinkhorn):
```

Comment:
isn't this supposed to inherite from `WasSolver`? It's not clear either why there should be a link to `Sinkhorn` (I would understand `LRSinkhorn`)

Reply:
For now, I think it makes sense, since if we were to use `WasSolver` with `LRSinkhornSolver`, we'd have:

- 3 loops (outermost GW loop, outer loop in LR Sinkhorn, Dykstra loop)
- `LRSinkhornSolver` would have to be modified to accept a quadratic problem (or have a specific inner solver for unbalanced LR GW, which is more complicated than this)

We definitely need a  abstraction for the solvers, this is just temporary; will come up with a better solution to this in the future.

I also think that our current LR GW (balanced) is a bit wrong when it comes to the convergence criterion (we exit the outermost GW loop when the successive costs are close; in the paper it should exist when the errors between succcessive `Q/R/g` are close [the convergence criterion for the linearized objective]).
This will be fixed in a future PR once the LR GW (balanced and unbalanced) are unified in this class.

Reply:
Conceptually, I am not sure exactly what we're inheriting from `Sinkhorn`? It seems `Sinkhorn` is used as a simpler type of `WassersteinSolver`?